### DISCUSSION TITLE
Removing first element in array of arrays causes error

### DISCUSSION THREAD

**Discussion by uxm** [livewire/livewire#3104] [7. Bugs (archived)]
I'm experiencing an error when I try to remove certain elements in an array of arrays.

The error looks like this in the console:
Uncaught (in promise) DOMException: String contains an invalid character livewire.js:13

At that point, the element removed will become a duplicate of another element, and basically adjusting the array after that won't work. It works just fine... sometimes. It seems to work consistently when removing the last element, but not others.

I have tried many ways to try to workaround this issue without success.

The data is being printed out in blade like so:
@foreach($stuff as $ele)
  {{ $ele['name'] }}
  {{ $ele['quantity'] }}
@endforeach

I'm on the most recent version of livewire.

**Comment by joshhanley** [comment]
@uxm can you please share some working code or a repo that demonstrates the issue?

**Comment by uxm** [comment]
Okay, so I have finally narrowed it down to where the error is.

    @foreach($foods as $food)
      {{ $ele['name'] }}
      {{ $ele['quantity'] }}
      <button @click="adjustFoodQuantityName='{{ $food['name'] }}'">Adjust Quantity</button>
    @endforeach`

The issue is with the `{{ $food['name'] }}` inside the `alpinejs` `@click` code. If this part is removed, then the issue doesn't happen. Essentially, this button is setting the value of an `input` field with alpine.

I have figured out how to workaround the issue by using `wire:click` to set the input value, such as with: `wire:click="adjustFoodQuantityName('{{ $food['name'] }}')"` however this is non-optimal since it requires a delay. With alpine the result is near instant.

Any way I can achieve this via Alpine?

If you'd like, I can post sample code to be able to replicate the issue.

**Comment by austenc** [comment]
@uxm it would be helpful if we could see more context, such as the Alpine data definition. You can certainly set properties in Alpine like you mentioned: `@click="adjustFoodQuantityName='{{ $food['name'] }}'"`

Perhaps one of your foods has a character in it which is throwing off the encoding for alpine?

If you can provide an example repo it will be easier for us to take a look.

**Comment by uxm** [comment]
@austenc The issue occurs when removing an element from the array, such as `unset($this->foods[$key]);` in the livewire code. I can post a bunch of code I created for recreating the issue.

**Comment by joshhanley** [comment]
@uxm yeah, can you please provide that code in a repo that recreates the issue? That'd be great, then we can just pull it down and investigate. Thanks!

**Comment by uxm** [comment]
@joshhanley Would that be a full laravel app, or just the files that are needed? `Foods.php`, etc.

**Comment by joshhanley** [comment]
@uxm ideally a full app, so we can just composer install, and then it's running 🙂 

Or you could try replicating the issue on https://laravelplayground.com/

**Comment by uxm** [comment]
@joshhanley Here you go:
https://github.com/uxm/samplefoodslistlivewirerepo

Steps to reproduce:
add multiple foods, of unique names.
delete the first food

example: use name: 1, quantity: 1, and name: 2, quantity: 2
then delete the first food, the first item will not delete, but rather be filled with name 2 quantity 2 (so there are two that say name 2 quantity 2)
![before-pressing-delete](https://user-images.githubusercontent.com/1063166/109557269-dd8b0480-7aa5-11eb-8f7f-b9a4917f8be6.png)
![after-pressing-delete](https://user-images.githubusercontent.com/1063166/109557280-dfed5e80-7aa5-11eb-808c-0c5999dc0d77.png)
![error](https://user-images.githubusercontent.com/1063166/109557283-e1b72200-7aa5-11eb-820f-5b4b656974bb.png)

**Comment by claytonrcarter** [comment]
Hi. Here's an as-simple-as-possible reproduction: https://laravelplayground.com/#/snippets/7c847999-8e61-4ed9-bc85-396db3494507

The issue is that `Element.setAttribute()` will throw a DOMException if the attribute name includes an `@` (and possibly other characters, but that's the one that brought me here). This is an issue if you have an element with Vue or Alpine handler that is changed in Blade and then needs to be morphed into the existing dom. So this code won't work `@click="doSomething({{ $phpValue }})` but this will `x-on:click="doSomething({{ $phpValue }})`. Also, this will work `x-data="{jsValue: {{ $phpValue }}}" @click="doSomething(jsValue)`.

The code that's causing the Exception is in `morphAttrs`, here: https://github.com/livewire/livewire/blob/b647ac63378ea02bdac975f70a4e57aea84bda0b/js/dom/morphdom/morphAttrs.js#L37

Perhaps this is a known issue or limitation or such, but it caught me as I'm trying to learn my way through this.

**Comment by uxm** [comment]
I tried editing the code, based on what @claytonrcarter said, from `@click` to `x-on:click` and it appears to resolve the issue. So the issue does seem to have to do with using `@` instead of `x-on`. Hopefully that helps

Within the sample repo I created, it would look like changing this:
`echo '<button style="font-size: .8rem;" @click="showAdjustFoodQuantity = !showAdjustFoodQuantity, adjustFoodQuantityName=\'' . $food['name'] . '\'" class="text-lg btn btn-primary">Adjust Quantity</button>';`

To:
`echo '<button style="font-size: .8rem;" x-on:click="showAdjustFoodQuantity = !showAdjustFoodQuantity, adjustFoodQuantityName=\'' . $food['name'] . '\'" class="text-lg btn btn-primary">Adjust Quantity</button>';`

**Comment by joshhanley** [comment]
@claytonrcarter using your third example in the playground provided, the issue is you don't have any `wire:key`.

I have added `wire:key` to your example here and it now works fine
https://laravelplayground.com/#/snippets/3c500959-eea2-49fd-9e1e-5ad385f33164

The reason is morphdom has no way to manipulate already assigned click handlers, so instead using `wire:key` tells morphdom if they are the same element or group of elements, and if so retain the click handlers. If not, it throws the elements out and creates them fresh, where Alpine can then take over and register the handlers.

@uxm having a quick look at your repo, you need to do the same, add a `wire:key="food-{{ $food[id] }}` onto the first div inside the `@foreach` loop, and see if that fixes it.

Let me know if it doesn't and I will pull it down and have a play.

Hope this helps!

**Comment by joleenshook** [comment]
I had this problem with a draggable group (using laravel-sortable) as soon as I put a little alpinejs lightbox script on an element inside of it.  The div had a child div that had a morphmany relationship in it (a group of files displayed as thumbnails).

**Oddly, all I needed to do to make it work again was change @click to x-on-click?**

I already had wire:key on the draggable div, and it would drag fine but error out with `Uncaught (in promise) DOMException: String contains an invalid character livewire.js:13.`  Therefore the function that renumbered the steps did not execute.

Here is the relevant code (with the change already made) in case it helps anyone resolve the issue.

        <div class="px-5 py-5 space-y-0">
            <div class="table">
                <div class="table-row-group" wire:sortable="updateStepOrder">
                    @foreach($steps as $step)
                        <div wire:sortable.item="{{$step->id}}" wire:key="{{$step->id}}"
                             class="table-row mb-5 flex border border-gray-300"
                             draggable="true">
                            <div
                                class="table-cell px-3 py-2 whitespace-nowrap border border-gray-300">
                                <x-icon-button wire:sortable.handle icon="fal fa-grip-vertical"></x-icon-button>
                                Step&nbsp;{{$step->order ?? ''}}
                            </div>
                            <div
                                class="table-cell border-b border-t border-r border-gray-300 whitespace-nowrap px-3 py-2">
                                <x-icon-button wire:click.prevent="editStep({{$step->id}})"
                                               icon="fal fa-pencil"></x-icon-button>
                                <x-icon-button wire:click.prevent="confirmDelete({{$step->id}})"
                                               icon="fal fa-trash-alt"></x-icon-button>
                            </div>
                            <div class="table-cell px-3 py-2 border-b border-t border-r border-gray-300"
                                 style="width:99%">
                                <div>
                                    {!! nl2br($step->content) ?? '' !!}
                                </div>
                                <div class="flex justify-items-start flex-row flex-wrap ">
                                    @if(!empty($step->files))
                                        @foreach ($step->files as $file)
                                        <div class="m-5">
                                            @if(strpos($file->filetype,'image') === 0)
                                                <a x-on:click="$dispatch('img-modal', {  imgModalSrc: '{{ $file->filepath ?? "" }}', imgModalDesc: '{{$step->order ?? ""}}' })"
                                                   class="cursor-pointer" >
                                                    <img class="object-cover object-center w-32 max-h-24" alt="{{$step->order ?? ''}} "
                                                         src="{{$file->filepath ?? asset('images/placeholder.jpg')}}"/>
                                                </a>
                                            @else
                                                <i class="fal fa-file-pdf text-3xl"></i>
                                            @endif
                                        </div>
                                        @endforeach
                                    @endif
                                </div>
                            </div>
                        </div>
                    @endforeach
                </div>
            </div>
        </div>
